docs/compilers/CSharp/Nullability Public API Design Notes.md

```diff
@@ -0,0 +1,241 @@
+When a class is declared, it doesn't have a nullability.
+It's just an `INamedTypeSymbol`.
+It may have constraints that have nullabilities, but the top-level nullability is not a question that makes any sense.
+However, in order to not break analyzers on nullable-unaware code, we will have to make the default nullability `Unknown`, as old C# code will have this as the nullability when `GetTypeInfo` is called on an expression.
+This would further exacerbate issue #1.
+3. This causes massive explosion in type heirarchies.
+If a user got `string?` and called `GetDeclaredMembers()`, the parent of each of these members must point back to `string?`, and not `string` or `string~`.
+
+For these reasons, we have decided not take this approach.
+
+### `GetSingleStatementSemanticModel`
+
+Instead of an overload on `Compilation` to get a `SemanticModel` without nullability info calculated, we had a `GetSingleStatementSemanticModel()` method on `SemanticModel`, that would get a `SemanticModel` that shared the `BoundNode` cache with the existing `SemanticModel` and did not calculate nullability information.
+This was rejected for being too complicated a solution, and for not being very future-proof if we were to add more analyses that users would want to opt out of.
+
+### `GetNullabilityAwareSemanticModel`
+
+This alternative proposal flipped the `GetSingleStatementSemanticModel` proposal on its head: instead of `SemanticModel` calculating nullability by default, it would only calculate when requested.
+We decided against this because we believe that, except for specific cases where perf is an issue, the savings gained from not calculating nullability would be moot as IDE features or analyzers will have already requested nullability anyway, and the time savings would be unused.
+Therefore, we are not taking this approach.
+
+### `GetNullabilityInfo`
+
+An early alternative proposal had a parallel API to `GetTypeInfo()`, `GetNullabilityInfo()`, that would return a structure similar to `TypeInfo` that contained nullability information.
+We decided that this would provide no benefits: because of nested nullability, we would still need to run nullability analysis on calls to `GetTypeInfo`, and it would needlessly separate information that users will want to get access to in tandem.
+Additionally, we could not come up with a good set of APIs for `GetDeclaredSymbol` and `GetSymbolInfo` to go along with that API.
+
+### Lazily calcualted `Nullability` properties
+
+This alternative proposal added similar `Nullability` properties to the interfaces as the main proposal, but did not calculate them ahead of time.
+Rather, they would be calculated when requested by the user.
+The big problem with this is that suddenly symbols, `TypeInfo`, and `SymbolInfo` must suddenly start carrying around the context in which they were called to be able to lazily calculate these properties, which is a major breaking change.
```

Comment:
can you clarify how that would be a breaking change?  Wouldn't it be an internal detail that users wouldn't know about?

Reply:
TypeInfo is currently a light wrapper type around two ITypeSymbols. The IDE hangs onto these all the time, and they don't have any references back to the context (read, position in a source tree) they came from. In order to calculate nullability information, you must have a location in a source tree, so these structs would suddenly no longer be lightweight wrappers.

Reply:
> TypeInfo is currently a light wrapper type around two ITypeSymbols. The IDE hangs onto these all the time, and TypeInfos don't have any references back to the context (read, position in a source tree) they came from. In order to calculate nullability information, you must have a location in a source tree, so these structs would suddenly no longer be lightweight wrappers.

IDE does not hang onto TypeInfos all the time.  Can you point me to examples that you're thinking of?    The XXXInfo guys are almost always ephemeral.  We do sometimes hold onto *Symbols* for shortish periods of time.  But even those we try to let go of since they can keep compilations alive.

Reply:
@jasonmalinowski or @DustinCampbell? I believe it was one of you who made that point about `TypeInfo` in the meetings. Regardless of that, we don't have interest in making `TypeInfo`s suddenly need to contain context to allow them to lazily calculate nullability.